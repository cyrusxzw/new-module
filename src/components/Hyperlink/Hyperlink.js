import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { Icon } from '~/components/Icon';
import {
  checkIsExternalFromStyle,
  checkIsInlineFromStyle,
  getTargetType,
  hasIconFromStyle,
} from './Hyperlink.utils';
import styles from './Hyperlink.module.css';

const Hyperlink = forwardRef(
  (
    {
      children,
      className,
      dataTestRef,
      hasTargetInNewWindow,
      isAlternate,
      isDownload,
      onClick,
      style,
      textAlign,
      theme,
      title,
      url,
    },
    ref,
  ) => {
    const isInline = checkIsInlineFromStyle(style);
    const isExternal = checkIsExternalFromStyle(style);
    const hasIcon = hasIconFromStyle(style);
    const target = getTargetType(hasTargetInNewWindow);
    const classSet = cx(
      styles.base,
      { [styles.blockStyle]: !isInline },
      { [styles.hasIcon]: hasIcon },
      { [styles.hasExternalIcon]: isExternal && hasIcon },
      { [styles.alternate]: isAlternate },
      styles[textAlign],
      styles[theme],
      className,
    );

    let iconName = '';

    if (isExternal) {
      iconName = 'rightUpArrow';
    } else {
      iconName = 'rightArrow';
    }

    return (
      <a
        className={classSet}
        data-test-ref={dataTestRef}
        download={isDownload}
        href={url}
        onClick={onClick}
        ref={ref}
        target={target}
        title={title}
      >
        {children}
        {hasIcon && (
          <>
            {` `}
            <i aria-hidden="true" className={styles.icon}>
              <Icon height={12} name={iconName} theme={theme} width={12} />
            </i>
          </>
        )}
      </a>
    );
  },
);

Hyperlink.propTypes = {
  children: PropTypes.any.isRequired,
  className: PropTypes.string,
  dataTestRef: PropTypes.string,
  hasTargetInNewWindow: PropTypes.bool,
  id: PropTypes.string,
  isAlternate: PropTypes.bool,
  isDownload: PropTypes.bool,
  onClick: PropTypes.func,
  style: PropTypes.oneOf([
    'External Button Link',
    'External No Icon Button Link',
    'External No Icon Link',
    'External Text Link',
    'Internal Button Link',
    'Internal No Icon Button Link',
    'Internal No Icon Link',
    'Internal Text Link',
  ]),
  textAlign: PropTypes.oneOf(['center', 'left', 'right']),
  theme: PropTypes.oneOf(['dark', 'light']),
  title: PropTypes.string,
  type: PropTypes.oneOf(['Relative', 'Absolute']),
  url: PropTypes.string.isRequired,
};

Hyperlink.defaultProps = {
  children: undefined,
  className: undefined,
  dataTestRef: undefined,
  hasTargetInNewWindow: false,
  id: undefined,
  isAlternate: undefined,
  isDownload: undefined,
  onClick: undefined,
  style: 'Internal No Icon Link',
  textAlign: 'left',
  theme: 'dark',
  title: undefined,
  type: undefined,
  url: undefined,
};

export { Hyperlink };
