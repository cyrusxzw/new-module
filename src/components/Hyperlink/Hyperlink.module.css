@import 'styles/index';

.base {
  position: relative;
  display: inline-block;
  font-family: var(--font-family-sans-serif-regular);
  font-size: inherit;
  font-weight: var(--font-weight-normal);
  outline: none;
  text-decoration: none;

  &,
  & * {
    box-sizing: border-box;
  }
}

.icon {
  display: inline-block;

  svg {
    width: 12px;
    max-height: 12px;

    path {
      transition: fill 150ms var(--easing-ease-in-out-cubic);
    }
  }
}

.blockStyle {
  display: inline-block;
  min-width: 210px;
  max-width: 300px;
  height: 60px;
  padding: 19px 23px;
  border-width: 1px;
  border-style: solid;
  font-family: var(--font-family-sans-serif-medium);
  font-size: rem(14px);
  line-height: 1.4;
  transition: all 200ms var(--easing-ease-in-out-cubic);

  @media (--viewport-xs-sm-only) {
    width: 100%;
    max-width: 100%;
  }

  &.hasIcon {
    padding-right: 55px;
  }

  .icon {
    position: absolute;
    top: calc(50% - 6px);
    right: 23px;
    font-size: 0;
    line-height: 1;
  }
}

.blockStyle.hasExternalIcon .icon {
  top: calc(50% - 5px);
}

.dark {
  @extend %font-theme-dark-link;

  .icon.icon {
    svg {
      path {
        fill: var(--color-dark-background);
      }
    }
  }

  &.blockStyle {
    border-color: var(--color-grey-rgba-20);

    .icon {
      svg {
        path {
          fill: var(--color-dark-background);
        }
      }
    }

    &:hover,
    &:focus {
      border-color: var(--color-dark-background);
      background-color: var(--color-dark-background);
      color: var(--color-light-link);

      .icon {
        svg {
          path {
            fill: var(--color-light-link);
          }
        }
      }
    }

    &.alternate {
      border-color: var(--color-dark-background);
      background-color: var(--color-dark-background);
      color: var(--color-light-link-hover);

      &:hover,
      &:focus {
        border-color: var(--color-black);
        background-color: var(--color-black);
      }
    }
  }
}

.light.light {
  @extend %font-theme-light-link;

  .icon.icon {
    svg {
      path {
        fill: var(--color-white);
      }
    }
  }

  &.blockStyle {
    border: 1px solid var(--color-white);

    .icon {
      svg {
        path {
          fill: var(--color-white);
        }
      }
    }

    &:focus,
    &:hover {
      background-color: var(--color-white);
      color: var(--color-dark-copy);

      .icon {
        svg {
          path {
            fill: var(--color-dark-copy);
          }
        }
      }
    }

    &.alternate {
      border-color: var(--color-light-link-hover);
      background-color: var(--color-light-link-hover);
      color: var(--color-dark-copy);

      &:hover,
      &:focus {
        border-color: var(--color-white);
        background-color: var(--color-white);
      }
    }
  }
}

.center {
  justify-content: center;
  text-align: center;
}

.left {
  justify-content: left;
  text-align: left;
}

.right {
  justify-content: right;
  text-align: right;
}
