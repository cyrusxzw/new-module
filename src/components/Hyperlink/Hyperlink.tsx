import React, { forwardRef } from 'react';
import cx from 'classnames';
import { Icon } from '~/components/Icon';
import {
  checkIsExternalFromStyle,
  checkIsInlineFromStyle,
  getTargetType,
  hasIconFromStyle,
} from './Hyperlink.utils';
import type { HyperlinkProps } from './Hyperlink.types';
import type { SvgName } from '../Icon/Icon.types';
import styles from './Hyperlink.module.css';

const Hyperlink = forwardRef<HTMLAnchorElement, HyperlinkProps>(
  (
    {
      children,
      className,
      dataTestId,
      dataTestRef,
      hasTargetInNewWindow = false,
      isAlternate,
      isDownload,
      kind,
      onClick,
      onMouseEnter,
      style,
      tabIndex,
      textAlign = 'left',
      theme = 'dark',
      title,
      url,
    },
    ref,
  ) => {
    const linkStyle = style || kind || 'Internal No Icon Link';
    const isInline = checkIsInlineFromStyle(linkStyle);
    const isExternal = checkIsExternalFromStyle(linkStyle);
    const hasIcon = hasIconFromStyle(linkStyle);
    const target = getTargetType(hasTargetInNewWindow);
    const classSet = cx(
      styles.base,
      { [styles.blockStyle]: !isInline },
      { [styles.hasIcon]: hasIcon },
      { [styles.hasExternalIcon]: isExternal && hasIcon },
      { [styles.alternate]: isAlternate },
      styles[textAlign],
      styles[theme],
      className,
    );

    let iconName: SvgName = '';

    if (isExternal) {
      iconName = 'rightUpArrow';
    } else {
      iconName = 'rightArrow';
    }

    return (
      <a
        className={classSet}
        data-test-id={dataTestId}
        data-test-ref={dataTestRef}
        download={isDownload}
        href={url}
        onClick={onClick}
        onMouseEnter={onMouseEnter}
        ref={ref}
        tabIndex={tabIndex}
        target={target}
        title={title}
      >
        {children}
        {hasIcon && (
          <>
            {` `}
            <i aria-hidden="true" className={styles.icon}>
              <Icon height={12} name={iconName} theme={theme} width={12} />
            </i>
          </>
        )}
      </a>
    );
  },
);

export { Hyperlink };
