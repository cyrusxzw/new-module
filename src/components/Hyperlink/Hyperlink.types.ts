import type { MouseEvent, ReactNode } from 'react';
import type { Themes } from '~/types';

type TextAlign = 'center' | 'left' | 'right';
type LinkStyle =
  | 'External Button Link'
  | 'External No Icon Button Link'
  | 'External No Icon Link'
  | 'External Text Link'
  | 'Internal Button Link'
  | 'Internal No Icon Button Link'
  | 'Internal No Icon Link'
  | 'Internal Text Link';

type HyperlinkProps = {
  children: ReactNode;
  className?: string;
  dataTestId?: string;
  dataTestRef?: string;
  hasTargetInNewWindow?: boolean;
  id?: string;
  isAlternate?: boolean;
  isDownload?: boolean;
  kind?: LinkStyle;
  onClick?: (event: MouseEvent) => void;
  /** @deprecated use `kind` instead */
  style?: LinkStyle;
  tabIndex?: number;
  textAlign?: TextAlign;
  theme?: Themes;
  title?: string;
  url: string;
};

export type { HyperlinkProps, LinkStyle };
