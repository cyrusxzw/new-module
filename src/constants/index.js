import * as BREAKPOINTS from './breakpoints.js';
import * as COLORS from './colors.js';
import * as GOOGLE_MAPS from './google-maps.js';
import * as HEADING from './heading.js';
import * as HYPERLINK_STYLE_TYPES from './hyperlink-style-types.js';
import * as STORES from './stores.js';
import * as TEST_REFS from './test_refs.js';
import * as TRANSITIONS from './transitions.js';
import * as HTML from './html.js';
import * as KEYBOARD_CODES from './keyboard-codes.js';
import * as LABELS from './labels.js';

export { BREAKPOINTS };
export { COLORS };
export { GOOGLE_MAPS };
export { HEADING };
export { HYPERLINK_STYLE_TYPES };
export { STORES };
export { TEST_REFS };
export { TRANSITIONS };
export { HTML };
export { KEYBOARD_CODES };
export { LABELS };
