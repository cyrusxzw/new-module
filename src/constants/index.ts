import { BREAKPOINTS } from './breakpoints';
import * as COLORS from './colors.js';
import * as GOOGLE_MAPS from './google-maps.js';
import * as HEADING from './heading.js';
import * as HTML from './html.js';
import * as HYPERLINK_STYLE_TYPES from './hyperlink-style-types.js';
import * as KEYBOARD_CODES from './keyboard-codes.js';
import * as LABELS from './labels.js';
import * as STORES from './stores.js';
import * as TEST_REFS from './test_refs.js';
import * as TRANSITIONS from './transitions.js';
import { VIEWPORTS } from './viewports';

export { BREAKPOINTS };
export { COLORS };
export { GOOGLE_MAPS };
export { HEADING };
export { HTML };
export { HYPERLINK_STYLE_TYPES };
export { KEYBOARD_CODES };
export { LABELS };
export { STORES };
export { TEST_REFS };
export { TRANSITIONS };
export { VIEWPORTS };
