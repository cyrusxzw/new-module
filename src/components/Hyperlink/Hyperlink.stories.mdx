import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import { Hyperlink } from './Hyperlink';

<Meta title="Components/Hyperlink" component={Hyperlink} />

# Hyperlink

<Canvas>
  <Story name="Base component">
    <Hyperlink
      isAlternate={knobs.boolean('isAlternate', false, 'Presentation')}
      kind={knobs.select(
        'kind',
        [
          'External Button Link',
          'External No Icon Button Link',
          'External No Icon Link',
          'External Text Link',
          'Internal Button Link',
          'Internal No Icon Button Link',
          'Internal No Icon Link',
          'Internal Text Link',
        ],
        'Internal Text Link',
        'Presentation',
      )}
      textAlign={knobs.select(
        'textAlign',
        ['left', 'center', 'right'],
        'center',
        'Presentation',
      )}
      theme={knobs.select('theme', ['dark', 'light'], 'dark', 'Presentation')}
      title="Hyperlink Title"
      url={knobs.text('url', '/', 'Copy')}
    >
      {knobs.text('children', 'Lorem ipsum', 'Copy')}
    </Hyperlink>
  </Story>
</Canvas>

## API

<ArgsTable of={Hyperlink} />
