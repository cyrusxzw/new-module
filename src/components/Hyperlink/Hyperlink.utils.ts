import includes from 'lodash/includes';
import { HYPERLINK_STYLE_TYPES } from '~/constants';
import type { LinkStyle } from './Hyperlink.types';

const {
  EXTERNAL_BUTTON_LINK,
  EXTERNAL_NO_ICON_BUTTON_LINK,
  EXTERNAL_NO_ICON_TEXT_LINK,
  EXTERNAL_TEXT_LINK,
  INTERNAL_BUTTON_LINK,
  INTERNAL_NO_ICON_TEXT_LINK,
  INTERNAL_TEXT_LINK,
} = HYPERLINK_STYLE_TYPES;

const checkIsInlineFromStyle = (style: LinkStyle): boolean => {
  const styleRepresentsInline = includes(
    [
      EXTERNAL_NO_ICON_TEXT_LINK,
      EXTERNAL_TEXT_LINK,
      INTERNAL_NO_ICON_TEXT_LINK,
      INTERNAL_TEXT_LINK,
    ],
    style,
  );

  return styleRepresentsInline ? true : false;
};

const checkIsExternalFromStyle = (style: LinkStyle): boolean => {
  const styleRepresentsAbsoluteLink = includes(
    [
      EXTERNAL_BUTTON_LINK,
      EXTERNAL_NO_ICON_BUTTON_LINK,
      EXTERNAL_NO_ICON_TEXT_LINK,
      EXTERNAL_TEXT_LINK,
    ],
    style,
  );

  return styleRepresentsAbsoluteLink ? true : false;
};

const hasIconFromStyle = (style: LinkStyle): boolean =>
  includes(
    [
      EXTERNAL_BUTTON_LINK,
      EXTERNAL_TEXT_LINK,
      INTERNAL_BUTTON_LINK,
      INTERNAL_TEXT_LINK,
    ],
    style,
  )
    ? true
    : false;

const getTargetType = (openInANewWindow: boolean): string =>
  openInANewWindow ? '_blank' : '_self';

export {
  checkIsExternalFromStyle,
  checkIsInlineFromStyle,
  getTargetType,
  hasIconFromStyle,
};
