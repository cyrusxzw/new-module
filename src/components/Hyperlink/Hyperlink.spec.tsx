import React from 'react';
import { render } from '@testing-library/react';
import { HYPERLINK_STYLE_TYPES } from '~/constants';
import { Hyperlink } from './Hyperlink';
import {
  checkIsExternalFromStyle,
  checkIsInlineFromStyle,
  getTargetType,
  hasIconFromStyle,
} from './Hyperlink.utils';

const {
  EXTERNAL_BUTTON_LINK,
  EXTERNAL_TEXT_LINK,
  EXTERNAL_NO_ICON_TEXT_LINK,
  INTERNAL_TEXT_LINK,
} = HYPERLINK_STYLE_TYPES;

describe('<Hyperlink />', () => {
  it('should render base component correctly', () => {
    const { container } = render(
      <Hyperlink url="https://aesop.com">Aesop</Hyperlink>,
    );

    expect(container).toMatchSnapshot();
  });
});

describe('Hyperlink.utils.checkIsInlineFromStyle', () => {
  it('should return false if `style` prop is `External Button Link`', () => {
    const actual = checkIsInlineFromStyle(EXTERNAL_BUTTON_LINK);

    expect(actual).toBe(false);
  });

  it('should return true if `style` prop is `External Text Link`', () => {
    const actual = checkIsInlineFromStyle(EXTERNAL_TEXT_LINK);

    expect(actual).toBe(true);
  });
});

describe('Hyperlink.utils.checkIsExternalFromStyle', () => {
  it('should return true if `style` prop is `External Button Link`', () => {
    const actual = checkIsExternalFromStyle(EXTERNAL_BUTTON_LINK);

    expect(actual).toBe(true);
  });

  it('should return false if `style` prop is `Internal Text Link`', () => {
    const actual = checkIsExternalFromStyle(INTERNAL_TEXT_LINK);

    expect(actual).toBe(false);
  });
});

// describe('Hyperlink.utils.getButtonPropsFromStyle', () => {
//   it('should return obect with `to` property if `style` prop is `Internal Text Link`', () => {
//     const actual = getButtonPropsFromStyle(INTERNAL_BUTTON_LINK, 'about');
//
//     expect(actual).toMatchObject({ to: 'about' });
//   });
//
//   it('should return obect with `href` property if `style` prop is `External Text Link`', () => {
//     const actual = getButtonPropsFromStyle(
//       EXTERNAL_TEXT_LINK,
//       'https://aesop.com',
//     );
//
//     expect(actual).toMatchObject({ href: 'https://aesop.com' });
//   });
// });

describe('Hyperlink.utils.getTargetType', () => {
  it('should return `_self` if `openInANewWindow` prop is passed as `falsey`', () => {
    const actual = getTargetType(false);

    expect(actual).toBe('_self');
  });

  it('should return `_blank` if `openInANewWindow` prop is passed as `true`', () => {
    const actual = getTargetType(true);

    expect(actual).toBe('_blank');
  });
});

describe('Hyperlink.utils.hasIconFromStyle', () => {
  it('should return `true` if `style` prop is passed as anything other than `External No Icon Link` or `No Icon Link`', () => {
    const actual = hasIconFromStyle(EXTERNAL_BUTTON_LINK);

    expect(actual).toBe(true);
  });

  it('should return `false` if `style` prop is passed as `External No Icon Link` or `No Icon Link`', () => {
    const actual = hasIconFromStyle(EXTERNAL_NO_ICON_TEXT_LINK);

    expect(actual).toBe(false);
  });
});
