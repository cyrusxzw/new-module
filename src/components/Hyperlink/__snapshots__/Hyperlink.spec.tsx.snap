// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Hyperlink /> should render base component correctly 1`] = `
<div>
  <a
    class="base left dark"
    href="https://aesop.com"
    target="_self"
  >
    Aesop
  </a>
</div>
`;
